rs/rosetta-api/ledger_canister_blocks_synchronizer/src/blocks.rs

```diff
@@ -1403,89 +1391,165 @@ impl Blocks {
+        // This way we only load 100000 blocks into memory at a time
+        const ROSETTA_BLOCKS_PROCESSED_PER_ROUND: u64 = 100000;
+        let mut hashed_block_idx_interval = std::ops::RangeInclusive::new(
+            *block_indices.start(),
+            std::cmp::min(
+                *block_indices.end(),
+                block_indices
+                    .start()
+                    .saturating_add(ROSETTA_BLOCKS_PROCESSED_PER_ROUND),
+            ),
+        );
+
+        // A helper function to store the current rosetta block and reset the fields
+        let mut flush_transactions =
+            |current_icp_block_timestamp: TimeStamp| -> Result<(), BlockStoreError> {
+                // We want the transactions to be sorted by their icp block index
+                let mut transactions: Vec<(u64, Transaction)> = icp_transactions_per_rosetta_block
+                    .take()
+                    .into_values()
+                    .collect();
+                transactions.sort_by(|(idx1, _), (idx2, _)| idx1.cmp(idx2));
+                let rosetta_block = RosettaBlock {
+                    index: *current_rosetta_block_index.borrow(),
+                    parent_hash: current_rosetta_block_parent_hash
+                        .borrow()
+                        .map(|hash| hash.into_bytes()),
+                    timestamp: *current_rosetta_block_timestamp.borrow(),
+                    transactions: transactions.into_iter().collect(),
                 };
-                transactions_hashes.clear();
+
+                // Set the fields to the next rosetta block
+                *current_rosetta_block_index.borrow_mut() += 1;
+                *num_rosetta_blocks_created.borrow_mut() += 1;
+                current_rosetta_block_parent_hash.replace(Some(HashOf::new(rosetta_block.hash())));
+                current_rosetta_block_timestamp.replace(current_icp_block_timestamp);
+                icp_transactions_per_rosetta_block.replace(HashMap::new());
+
+                self.store_rosetta_block(rosetta_block, &mut statement1, &mut statement2)?;
+                Ok(())
+            };
+
+        loop {
+            // We extract the icp block indices from the current interval
+            let hashed_blocks = database_access::read_hashed_blocks(
+                &mut statement3,
+                named_params! {
+                    ":start":hashed_block_idx_interval.start(),
+                    ":end":hashed_block_idx_interval.end()
+                },
+            )?;
+
+            for block in hashed_blocks.into_iter() {
+                let transaction = Block::decode(block.block)?.transaction;
+                let transaction_hash = transaction.hash();
+                if icp_transactions_per_rosetta_block
+                    .borrow()
+                    .contains_key(&transaction_hash)
+                    || block.timestamp != *current_rosetta_block_timestamp.borrow()
```

Comment:
I suppose that in practice, when `block.timestamp != *current_rosetta_block_timestamp.borrow()`, then `block.timestamp` will always be larger?

Reply:
Yes because timestamps are strictly monotonically increasing with icp block indices